core/tracker/tracker.go

```diff
@@ -134,11 +149,55 @@ func analyseDutyFailed(duty core.Duty, es []event) (bool, component, string) {
 // TODO(xenowits): Implement logic for reporting duties.
 func failedDutyReporter(core.Duty, bool, string, string) {}
 
-// analyseParticipation returns the share indexes of peers that participated in this duty.
-// TODO(dhruv): implement logic to analyse participation.
-//nolint:deadcode
-func analyseParticipation(core.Duty, []event) []int {
-	return nil
+// analyseParticipation returns a set of share indexes of participated peers corresponding to each DV public key.
+func analyseParticipation(events []event) map[core.PubKey]map[shareIdx]bool {
+	// Set of shareIdx of participated peers per DV.
+	resp := make(map[core.PubKey]map[shareIdx]bool)
+
+	eventsPerDV := make(map[core.PubKey][]event)
+	for _, e := range events {
+		eventsPerDV[e.pubkey] = append(eventsPerDV[e.pubkey], e)
+	}
+
+	for pubKey, dvEvents := range eventsPerDV {
+		resp[pubKey] = make(map[shareIdx]bool)
+		for _, e := range dvEvents {
+			// If we get a parSigDBInternal event, then the current node participated successfully.
+			// If we get a parSigEx event, then the corresponding peer with e.shareIdx participated successfully.
+			if e.component == parSigDBInternal || e.component == parSigEx {
+				resp[pubKey][e.shareIdx] = true
+			}
+		}
+	}
+
+	return resp
+}
+
+// newParticipationReporter returns a new participation reporter function which logs and instruments peer participation
+func newParticipationReporter(peers []p2p.Peer) func(context.Context, core.Duty, map[core.PubKey]map[shareIdx]bool, map[core.PubKey]map[shareIdx]bool) {
```

Comment:
internalise state my keeping last reported state inside this function, do not require it to kept outside